KIDNEY STONE PREdICTION

1. BASICS 

In [ ]:
import pandas as pd
import numpy as np
from fastbook import *

In [ ]:

!pip install -Uqq fastbook
import fastbook
fastbook.setup_book()

In [ ]:
from fastai.vision.all import *
path = Path('/content/gdrive/MyDrive/kidney')

To count the images

In [ ]:
path = Path('/content/gdrive/MyDrive/kidney')
for folder in path.glob('*'):
    print(f"{folder}: {len(list(folder.glob('*')))}")



/content/gdrive/MyDrive/kidney/Normal: 1366
/content/gdrive/MyDrive/kidney/Stone: 1366


Preprocessing

In [ ]:
import os
import numpy as np
from PIL import Image
from keras.preprocessing.image import ImageDataGenerator




In [ ]:
datagen = ImageDataGenerator(
    rescale=1./255,
    rotation_range=20,
    width_shift_range=0.2,
    height_shift_range=0.2,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,
    fill_mode='nearest')


preprocessing


In [ ]:
from keras.preprocessing.image import ImageDataGenerator

# Define the ImageDataGenerator for preprocessing
train_datagen = ImageDataGenerator(
    rescale=1./255, 
    rotation_range=20,
    width_shift_range=0.2,
    height_shift_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,
    vertical_flip=True,
    fill_mode='nearest')


img_height = 224
img_width = 224
# Load the images from directory and apply data augmentation
train_generator = train_datagen.flow_from_directory(
    '/content/gdrive/MyDrive/kidney',
    target_size=(img_height, img_width),
    batch_size=batch_size,
    class_mode='binary')


Found 2733 images belonging to 2 classes.


In [ ]:
import os
print(os.listdir('/content/gdrive/MyDrive/kidney/Normal'))


['Normal- (203).jpg', 'Normal- (1932).jpg', 'Normal- (2016).jpg', 'Normal- (1981).jpg', 'Normal- (2023).jpg', 'Normal- (2041).jpg', 'Normal- (2002).jpg', 'Normal- (1982).jpg', 'Normal- (1931).jpg', 'Normal- (1984).jpg', 'Normal- (1977).jpg', 'Normal- (1934).jpg', 'Normal- (2030).jpg', 'Normal- (1949).jpg', 'Normal- (1927).jpg', 'Normal- (1966).jpg', 'Normal- (2018).jpg', 'Normal- (2032).jpg', 'Normal- (1999).jpg', 'Normal- (1924).jpg', 'Normal- (1965).jpg', 'Normal- (1960).jpg', 'Normal- (2010).jpg', 'Normal- (2039).jpg', 'Normal- (1972).jpg', 'Normal- (1947).jpg', 'Normal- (2036).jpg', 'Normal- (2009).jpg', 'Normal- (1986).jpg', 'Normal- (1955).jpg', 'Normal- (1944).jpg', 'Normal- (2021).jpg', 'Normal- (2027).jpg', 'Normal- (1942).jpg', 'Normal- (1921).jpg', 'Normal- (2024).jpg', 'Normal- (1989).jpg', 'Normal- (2022).jpg', 'Normal- (1943).jpg', 'Normal- (1988).jpg', 'Normal- (1950).jpg', 'Normal- (1971).jpg', 'Normal- (196).jpg', 'Normal- (1937).jpg', 'Normal- (194).jpg', 'Normal- (20

SPLIT


In [ ]:
!pip install split-folders
import os
import splitfolders 
input_folder = "/content/gdrive/MyDrive/kidney"
output = "/content/gdrive/MyDrive/new kidney"  
splitfolders.ratio(input_folder, output, seed=1337, ratio=(.75, .25))

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/



Copying files: 0 files [00:00, ? files/s]
Copying files: 1 files [00:00,  3.17 files/s]
Copying files: 9 files [00:00, 24.54 files/s]
Copying files: 15 files [00:01,  9.41 files/s]
Copying files: 23 files [00:01, 16.38 files/s]
Copying files: 29 files [00:01, 21.72 files/s]
Copying files: 38 files [00:01, 31.76 files/s]
Copying files: 45 files [00:01, 38.13 files/s]
Copying files: 52 files [00:02, 43.96 files/s]
Copying files: 60 files [00:02, 51.17 files/s]
Copying files: 67 files [00:02, 53.73 files/s]
Copying files: 75 files [00:02, 60.01 files/s]
Copying files: 82 files [00:02, 58.56 files/s]
Copying files: 90 files [00:02, 63.76 files/s]
Copying files: 98 files [00:02, 64.55 files/s]
Copying files: 106 files [00:02, 67.03 files/s]
Copying files: 114 files [00:02, 68.39 files/s]
Copying files: 122 files [00:03, 64.24 files/s]
Copying files: 130 files [00:03, 67.48 files/s]
Copying files: 137 files [00:03, 62.87 files/s]
Copying files: 145 files [00:03, 67.11 files/s]
Copying files

MODEL

INCEPTION V3


In [ ]:
# importing necceasary libraries
from keras.models import Model
from keras.layers import Dense
from keras.optimizers import Adam
from keras.layers import GlobalAveragePooling2D
from keras.applications.inception_v3 import InceptionV3

img_height = 224
img_width = 224
num_classes = 2

# Load the InceptionV3 model without the top layers
base_model = InceptionV3(weights='imagenet', include_top=False, input_shape=(img_height, img_width, 3))

# Add custom top layers for our dataset
x = base_model.output
x = GlobalAveragePooling2D()(x)
x = Dense(1024, activation='relu')(x)
predictions = Dense(num_classes, activation='softmax')(x)

# Combine the base model and the custom top layers
model = Model(inputs=base_model.input, outputs=predictions)

# Compile the model with Adam optimizer and categorical crossentropy loss
model.compile(optimizer=Adam(lr=0.0001), loss='categorical_crossentropy', metrics=['accuracy'])



Copying files: 927 files [29:43,  1.92s/ files]
/usr/local/lib/python3.9/dist-packages/keras/optimizers/legacy/adam.py:117: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super().__init__(name, **kwargs)


In [ ]:
train_dir = '/content/gdrive/MyDrive/new kidney/train'
num_classes = len(os.listdir(train_dir))

In [ ]:
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, GlobalAveragePooling2D
from tensorflow.keras.applications.inception_v3 import InceptionV3, preprocess_input


In [ ]:
#hyperparameters
img_height = 224
img_width = 224
batch_size = 32
num_epochs = 10
num_classes = 2

In [ ]:

# Create data generators for train and validation sets
train_datagen = ImageDataGenerator(
    rescale=1./255,
    rotation_range=20,
    zoom_range=0.15,
    width_shift_range=0.2,
    height_shift_range=0.2,
    shear_range=0.15,
    horizontal_flip=True,
    fill_mode='nearest')

val_datagen = ImageDataGenerator(rescale=1./255)

In [ ]:
train_generator = train_datagen.flow_from_directory(
    '/content/gdrive/MyDrive/new kidney/train',
    target_size=(img_height, img_width),
    batch_size=batch_size,
    class_mode='categorical')

Found 2048 images belonging to 2 classes.


In [ ]:
val_generator = val_datagen.flow_from_directory(
    '/content/gdrive/MyDrive/new kidney/val',
    target_size=(img_height, img_width),
    batch_size=batch_size,
    class_mode='categorical')

Found 684 images belonging to 2 classes.


In [ ]:

# Load the InceptionV3 model without the top layers
base_model = InceptionV3(weights='imagenet', include_top=False, input_shape=(img_height, img_width, 3))


In [ ]:
# Add custom top layers for  dataset
x = base_model.output
x = GlobalAveragePooling2D()(x)
x = Dense(1024, activation='relu')(x)
predictions = Dense(num_classes, activation='softmax')(x)

In [ ]:
# Combine the base model and the custom top layers
model = tf.keras.models.Model(inputs=base_model.input, outputs=predictions)

# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])


In [ ]:
# Train the model
history = model.fit(
    train_generator,
    epochs=num_epochs,
    validation_data=val_generator,
    batch_size=batch_size,
    steps_per_epoch=train_generator.samples//batch_size,
    validation_steps=val_generator.samples//batch_size)

Epoch 1/10
64/64 [==============================] - 1403s 22s/step - loss: 0.3255 - accuracy: 0.8979 - val_loss: 23.5926 - val_accuracy: 0.5000
Epoch 2/10
64/64 [==============================] - 1355s 21s/step - loss: 0.0855 - accuracy: 0.9731 - val_loss: 4.3005 - val_accuracy: 0.7143
Epoch 3/10
64/64 [==============================] - 1360s 21s/step - loss: 0.0472 - accuracy: 0.9858 - val_loss: 2.4094 - val_accuracy: 0.8080
Epoch 4/10
64/64 [==============================] - 1406s 22s/step - loss: 0.0411 - accuracy: 0.9844 - val_loss: 0.6050 - val_accuracy: 0.8601
Epoch 5/10
64/64 [==============================] - 1403s 22s/step - loss: 0.0082 - accuracy: 0.9980 - val_loss: 0.0025 - val_accuracy: 1.0000
Epoch 6/10
21/64 [========>.....................] - ETA: 14:03 - loss: 2.3670e-04 - accuracy: 1.0000

SAVING THE MODEL

In [ ]:
model.save('/content/drive/MyDrive/kidneymodel/inception.h5')
+7

when testing on new image...from class normal

In [ ]:
from tensorflow import keras
model = keras.models.load_model('/content/drive/MyDrive/kidneymodel/inception.h5')

In [ ]:
from tensorflow.keras.preprocessing import image
import numpy as np
from tensorflow.keras.applications.inception_v3 import InceptionV3, preprocess_input

img_height = 224
img_width = 224

# Load the image
img = image.load_img('/content/drive/MyDrive/kidney/Normal/Normal- (1002).jpg', target_size=(img_height, img_width))

# Preprocess the image
img = image.img_to_array(img)
img = np.expand_dims(img, axis=0)
img = preprocess_input(img)

# Make predictions
preds = model.predict(img)

print(preds)


1/1 [==============================] - 0s 86ms/step
[[0.5074756]]


So, in this case, the predicted class for the input image '/content/drive/MyDrive/kidney/Normal/Normal- (1002).jpg' would be a normal kidney, since the predicted probability (0.5074756) is lesser than the threshold of 0.5.

In [ ]:
from tensorflow.keras.preprocessing import image
import numpy as np
from tensorflow.keras.applications.inception_v3 import InceptionV3, preprocess_input

img_height = 224
img_width = 224

# Load the image
img = image.load_img('/content/drive/MyDrive/kidney/Stone/Stone- (210).jpg', target_size=(img_height, img_width))

# Preprocess the image
img = image.img_to_array(img)
img = np.expand_dims(img, axis=0)
img = preprocess_input(img)

# Make predictions
preds = model.predict(img)

print(preds)


1/1 [==============================] - 0s 73ms/step
[[0.52406675]]


 the probability is 0.52, which is greater than 0.5, which means the model predicts that the input image contains a kidney stone.

output with printing stone or not

In [ ]:
img_height = 224
img_width = 224
img = image.load_img('/content/drive/MyDrive/kidney/Stone/Stone- (1016).jpg', target_size=(img_height, img_width))
img = image.img_to_array(img)
img = np.expand_dims(img, axis=0)
img = preprocess_input(img)
print(preds)
preds = model.predict(img)
threshold = 0.5
if preds > threshold:
    print("The input image contains a kidney stone.")
else:
    print("The input image contains a normal kidney.")


1/1 [==============================] - 0s 175ms/step
[[0.5065489]]
1/1 [==============================] - 0s 190ms/step
The input image contains a kidney stone.


In [ ]:
img_height = 224
img_width = 224
img = image.load_img('/content/drive/MyDrive/kidney/Normal/Normal- (1108).jpg', target_size=(img_height, img_width))
img = image.img_to_array(img)
img = np.expand_dims(img, axis=0)
img = preprocess_input(img)
preds = model.predict(img)
print(preds)
threshold = 0.5
if preds > threshold:
    print("The input image contains a kidney stone.")
else:
    print("The input image contains a normal kidney.")


1/1 [==============================] - 0s 165ms/step
[[0.47391114]]
The input image contains a normal kidney.


In [ ]:
img_height = 224
img_width = 224
img = image.load_img('/content/drive/MyDrive/kidney/Stone/Stone- (428).jpg', target_size=(img_height, img_width))
img = image.img_to_array(img)
img = np.expand_dims(img, axis=0)
img = preprocess_input(img)
preds = model.predict(img)
print(preds)
threshold = 0.5
if preds > threshold:
    print("The input image contains a kidney stone.")
else:
    print("The input image contains a normal kidney.")


1/1 [==============================] - 0s 127ms/step
[[0.50312525]]
The input image contains a kidney stone.
